services/headless-lms/src/regrading.rs

```diff
@@ -0,0 +1,421 @@
+            {
+                let not_ready_grading =
+                    models::gradings::new_grading(&mut *conn, &submission).await?;
+                models::submissions::set_grading_id(
+                    &mut *conn,
+                    not_ready_grading.id,
+                    submission.id,
+                )
+                .await?;
+                let entry = grading_futures
+                    .entry(exercise_task.exercise_type.clone())
+                    .or_default();
+
+                // make sure we aren't sending too many requests
+                let limit = usize::try_from(exercise_service.max_reprocessing_submissions_at_once)
+                    .unwrap_or_else(|_e| {
+                        tracing::error!(
+                            "{}: invalid max_reprocessing_submissions_at_once {}",
+                            exercise_service.name,
+                            exercise_service.max_reprocessing_submissions_at_once
+                        );
+                        usize::MAX
+                    });
+                if entry.len() < limit {
+                    let exercise_type = exercise_task.exercise_type.clone();
+                    let grading_future = models::gradings::send_grading_request(
+                        info,
+                        exercise_task,
+                        submission.clone(),
+                    )
+                    .map(|r| (regrading_submission, not_ready_grading, exercise, r));
+                    let entry = grading_futures.entry(exercise_type).or_default();
+                    entry.push(Box::pin(grading_future));
+                    if !sent_submission {
+                        sent_submission = true;
+                        models::regradings::set_grading_progress(
+                            &mut *conn,
+                            regrading_id,
+                            GradingProgress::Pending,
+                        )
+                        .await?;
+                    }
+                } else {
+                    // we can't send this submission right now, so mark the related regrading as incomplete
+                    incomplete_regradings.insert(regrading_id);
+                }
+            } else {
+                tracing::warn!(
+                    "No exercise service found for type {}",
+                    exercise_task.exercise_type
+                );
+            }
+        }
+    }
+
+    // wait for all the submissions to be completed
+    let mut grading_futures = grading_futures
+        .into_iter()
+        .map(|v| v.1)
+        .flatten()
+        .collect::<FuturesUnordered<_>>();
+    while let Some((regrading_submission, new_grading, exercise, grading_result)) =
+        grading_futures.next().await
+    {
+        // update regrading submission grading
+        let regrading_submission: RegradingSubmission = regrading_submission;
+        models::regrading_submissions::set_grading_after_regrading(
```

Comment:
If this is set and the grading (for this submission) behind is complete, we can skip above